### PR TITLE
Don't lint needless_borrow when matching `&&mut` by `&ref`

### LINKED ISSUE #1432: needless_borrow false-positive when destructuring `&&mut T`
```rust
let mut s = String::new();
match &&mut s {
    &ref a => {} // WARNING: this pattern creates a reference to a reference (clippy)
}
```

```rust
let mut s = String::new();
match &&mut s {
    &a => {} // ERROR: cannot move out of borrowed content
}
```

### PR DESCRIPTION
**Pull request by sinkuu**
Fixes #1432

### CODE REVIEW

File: tests/compile-fail/needless_borrow.rs [hunk 1]
```diff
@@ -33,3 +33,18 @@ trait Trait {}
 impl<'a> Trait for &'a str {}
 
 fn h(_: &Trait) {}
+
+#[allow(dead_code, unused_mut, match_ref_pats, unused_variables)]
+fn issue_1432() {
+    let mut s = String::new();
+    match & &mut s {
+        & ref a => {}
```
Comment (oli-obk) [inline tests/compile-fail/needless_borrow.rs]: we could suggest to change this to just `a`, or am I missing something?

Comment (sinkuu) [inline tests/compile-fail/needless_borrow.rs]: I was using `match` to exhibit the behavior. Updated to more realistic code.

Comment (oli-obk) [inline tests/compile-fail/needless_borrow.rs]: oh, that was fine, I meant the `&ref` pattern, I added an extra issue for it: https://github.com/Manishearth/rust-clippy/issues/1434

The new examples are great!

**Comment by oli-obk** [comment]
thanks!

**Pull request merged**